src/client/FSCrypt.cc

```diff
@@ -0,0 +1,1137 @@
+                      char *dest, int dest_len)
+{
+  char extract_buf[CEPH_CRYPTO_HMACSHA512_DIGESTSIZE];
+  int r = hkdf_extract(salt, salt_len,
+                       key, key_len,
+                       extract_buf, sizeof(extract_buf));
+  if (r < 0) {
+    return r;
+  }
+
+  int extract_len = r;
+
+//#define FSCRYPT_INFO_STR "fscrypt\x00\x01"
+#define FSCRYPT_INFO_STR "fscrypt\x00?"
+
+  char info_str[sizeof(FSCRYPT_INFO_STR) + nonce_len];
+
+  int len = sizeof(FSCRYPT_INFO_STR) - 1;
+  memcpy(info_str, FSCRYPT_INFO_STR, len);
+
+  info_str[len - 1] = hkdf_context;
+
+  if (nonce && nonce_len) {
+    memcpy(info_str + len, nonce, nonce_len);
+    len += nonce_len;
+  }
+
+  r =  hkdf_expand(extract_buf, extract_len,
+                   info_str, len,
+                   dest, dest_len);
+
+  return r;
+}
+
+static std::string hex_str(const void *p, int len)
+{
+  bufferlist bl;
+  bl.append_hole(len);
+  memcpy(bl.c_str(), p, len);
+  std::stringstream ss;
+  bl.hexdump(ss);
+  return ss.str();
+}
+
+std::ostream& operator<<(std::ostream& out, const ceph_fscrypt_key_identifier& kid) {
+  out << hex_str(kid.raw, sizeof(kid.raw));
+  return out;
+}
+
+int FSCryptKey::init(const char *k, int klen) {
+  int r = fscrypt_calc_hkdf(HKDF_CONTEXT_KEY_IDENTIFIER,
+                            nullptr, 0, /* nonce */
+                            nullptr, 0, /* salt */
+                            (const char *)k, klen,
+                            identifier.raw, sizeof(identifier.raw));
+  if (r < 0) {
+    return r;
+  }
+
+  key.append_hole(klen);
+  memcpy(key.c_str(), k, klen);
+
+  return 0;
+}
+
+int FSCryptKey::calc_hkdf(char ctx_identifier,
+                          const char *nonce, int nonce_len,
+                          char *result, int result_len) {
+  int r = fscrypt_calc_hkdf(ctx_identifier,
+                            nonce, nonce_len, /* nonce */
+                            nullptr, 0, /* salt */
+                            (const char *)key.c_str(), key.length(),
+                            result, result_len);
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+int ceph_fscrypt_key_identifier::init(const char *k, int klen) {
+  if (klen != sizeof(raw)) {
+    return -EINVAL;
+  }
+  memcpy(raw, k, klen);
+
+  return 0;
+}
+
+int ceph_fscrypt_key_identifier::init(const struct fscrypt_key_specifier& k) {
+  if (k.type != FSCRYPT_KEY_SPEC_TYPE_IDENTIFIER) {
+    return -EINVAL;
+  }
+
+  return init((const char *)k.u.identifier, sizeof(k.u.identifier));
+}
+
+bool ceph_fscrypt_key_identifier::operator<(const struct ceph_fscrypt_key_identifier& r) const {
+  return (memcmp(raw, r.raw, sizeof(raw)) < 0);
+}
+
+void FSCryptContext::generate_iv(uint64_t block_num, FSCryptIV& iv) const
+{
+  memset(&iv, 0, sizeof(iv));
+
+  // memcpy(iv.u.nonce, nonce, FSCRYPT_FILE_NONCE_SIZE);
+  iv.u.block_num = block_num;
+}
+
+void FSCryptContext::generate_new_nonce()
+{
+  cct->random()->get_bytes((char *)nonce, sizeof(nonce));
```

Comment:
double check this is using a good CPRNG

Reply:
It should be sufficient.

If `getentropy` is available that is used. If not available it is based off of `BCryptGenRandom` for Windows and then `/dev/urandom` for all else. `random()` is used in other places such as rgw, cephx etc.